nupic/research/frameworks/dendrites/routing/utils.py

```diff
@@ -103,3 +103,132 @@ def get_gating_context_weights(output_masks, context_vectors, num_dendrites):
+
+    :param model: a torch.nn.Module subclass that implements a dendrite module in
+                  addition to a linear feed-forward module, and takes both feedforward
+                  and context inputs to its `forward` method
+    :param loader: a torch dataloader that iterates over all train and test batches
+    :param optimizer: optimizer object used to train the model
+    :param device: device to use ('cpu' or 'cuda')
+    :param criterion: loss function to minimize
+    """
+    model.train()
+
+    for data_and_context, target in loader:
+
+        data_and_context = data_and_context.to(device)
+        target = target.to(device)
+
+        optimizer.zero_grad()
+        output = model(data_and_context)
+        loss = criterion(output, target)
+
+        loss.backward()
+        optimizer.step()
+
+
+def evaluate_dendrite_model(model, loader, device, criterion):
+    """
+    Evaluates a dendritic network on a specified criterion by iterating through all
+    batches in the given dataloader
+
+    :param model: a torch.nn.Module subclass that implements a dendrite module in
+                  addition to a linear feed-forward module, and takes both feedforward
+                  and context inputs to its `forward` method
+    :param loader: a torch dataloader that iterates over all train and test batches
+    :param device: device to use ('cpu' or 'cuda')
+    :param criterion: loss function to minimize
+    """
+    model.to(device)
+    model.eval()
+    loss = torch.tensor(0.0, device=device)
+
+    with torch.no_grad():
+        for data, context, target in loader:
+
+            data = data.to(device)
+            context = context.to(device)
+            target = target.to(device)
+
+            output = model(data, context)
+            loss += criterion(output, target, reduction="sum")
+
+    loss = loss.item()
+    return {"loss": loss}
+
+
+def evaluate_non_dendrite_model(model, loader, device, criterion):
+    """
+    Evaluates a regular network on a specified criterion by iterating through all
+    batches in the given dataloader
+
+    :param model: a torch.nn.Module subclass that implements a dendrite module in
+                  addition to a linear feed-forward module, and takes both feedforward
+                  and context inputs to its `forward` method
+    :param loader: a torch dataloader that iterates over all train and test batches
+    :param device: device to use ('cpu' or 'cuda')
+    :param criterion: loss function to minimize
+    """
+    model.to(device)
+    model.eval()
+    loss = torch.tensor(0.0, device=device)
+
+    with torch.no_grad():
+        for data_and_context, target in loader:
+
+            data_and_context = data_and_context.to(device)
+            target = target.to(device)
+
+            output = model(data_and_context)
+            loss += criterion(output, target, reduction="sum")
+
+            # Report mean absolute error for masked and unmasked indices separately
```

Comment:
What are these lines doing, and why is it different for non-dendrite models? There's a lot of duplicated code. Ideally the training and evaluation loops should use the exact same functions for both dendrite and non-dendrite models. The only difference is whether the context vectors are concatenated or not.

Reply:
Ok, we don't need all those metrics, so I just included mean absolute error (in addition to `criterion` provided as an arg). Also, I combined the train and evaluate functions to remove duplicate code.